modules/standard/CommDiagnostics.chpl

```diff
@@ -17,86 +17,269 @@
+   the operations for later retrieval.  The former gives more detailed
+   information but has much more overhead.  The latter has much less
+   overhead but only provides aggregate information.
+
+   All forms of statistics collection are done between pairs of function
+   calls that turn it on and off.  On-the-fly reporting across all
+   locales is done like this::
+
+     startVerboseComm();
+     // between start/stop calls, report comm ops initiated on any locale
+     stopVerboseComm();
+
+   On-the-fly reporting for a single locale is done like this::
+
+     startVerboseCommHere();
+     // between start/stop calls, report comm ops initiated on this locale
+     stopVerboseCommHere();
+
+   Aggregate statistics collection requires a few more calls.  One has
+   to retrieve the statistics after they are collected and, if they are
+   to be used again, reset the internal counters afterward.  Collecting
+   statistics across all locales is done like this::
+
+     // (optional) if we counted previously, reset the counters to zero
+     resetCommDiagnostics();
+     startCommDiagnostics();
+     // between start/stop calls, count comm ops initiated on any locale
+     stopCommDiagnostics();
+     // retrieve the counts and report the results
+     writeln(getCommDiagnostics());
+
+   Collecting statistics on just the calling locale is much the same.
+   Only the procedure names change::
+
+     // (optional) if we counted previously, reset the counters to zero
+     resetCommDiagnosticsHere();
+     startCommDiagnosticsHere();
+     // between start/stop calls, count comm ops initiated on this locale
+     stopCommDiagnosticsHere();
+     // retrieve the counts and report the results
+     writeln(getCommDiagnosticsHere());
+
+   The optional call to reset the counters is only needed when a program
+   collects communication statistics more than once.  In this case, the
+   counters have to be set back to zero before starting the second and
+   succeeding collection periods.  By far the most common situation is
+   that programs only collect the communication counters once per run,
+   in which case this step is not needed.
+
+   Note that the same internal mechanisms and counters are used for
+   counting on all locales and counting on just the calling locale, so
+   trying to do both at once may lead to surprising turn-of/turn-off
+   behavior and/or incorrect results.
+ */
 module CommDiagnostics
 {
   //
   // multi-locale diagnostics/debugging support
   //
-  
+
   // There should be a type like this declared in chpl-comm.h with a single
   // function that returns the C struct.  We're not doing it that way yet
   // due to some shortcomings in our extern records implementation.
   // Once that gets sorted out, we can turn this into an extern record,
   // and remove the 8 or so individual functions below that return the
   // various counters.
+
+  /* Aggregated communication operation counts.  This record type is
```

Comment:
It's unfortunate that we have to expose this internal type but I agree that we need to.  I think we should avoid talking about the runtime in end-user facing library documentation.